File: tools/rebuild_npc_spawns_317.js
const fs = require('fs');
const path = require('path');

const ROOT = 'd:/CodingProjects/OSRSRSPS';
const osrsNpcsPath = path.join(ROOT, 'osrs_npcs.json');
const npcDefsPath = path.join(ROOT, 'server/data/definitions/npc_defs.json');
const mapIndexPath = path.join(ROOT, 'server/data/clipping/map_index');
const outPath = path.join(ROOT, 'server/data/definitions/npc_spawns.json');

const toFacing = (direction) => {
  const sector = Math.floor((((Number(direction) || 0) + 128) % 2048) / 256);
  return {
    0: 'SOUTH', 1: 'SOUTH_WEST', 2: 'WEST', 3: 'NORTH_WEST',
    4: 'NORTH', 5: 'NORTH_EAST', 6: 'EAST', 7: 'SOUTH_EAST',
  }[sector] || 'SOUTH';
};

const regionIdForTile = (x, y) => ((x >> 6) << 8) | (y >> 6);

const defs = JSON.parse(fs.readFileSync(npcDefsPath, 'utf8'));
const validIds = new Set(defs.filter(d => d && Number.isInteger(d.id)).map(d => d.id));

const mapIndex = fs.readFileSync(mapIndexPath);
const count = mapIndex.readUInt16BE(0);
const validRegions = new Set();
for (let i = 0, off = 2; i < count && off + 5 < mapIndex.length; i++, off += 6) {
  const region = mapIndex.readUInt16BE(off);
  const landscape = mapIndex.readUInt16BE(off + 2);
  const objects = mapIndex.readUInt16BE(off + 4);
  if (landscape > 0 || objects > 0) validRegions.add(region);
}

const entries = JSON.parse(fs.readFileSync(osrsNpcsPath, 'utf8'));
const stats = { total: 0, kept: 0, skip_bad: 0, skip_id: 0, skip_region: 0, skip_coords: 0, skip_plane: 0, dupes: 0 };
const seen = new Set();
const out = [];

for (const e of entries) {
  stats.total++;
  const npcId = Number(e?.id), x = Number(e?.x), y = Number(e?.y), z = Number(e?.z ?? e?.p ?? 0);
  const radiusRaw = Number(e?.radius ?? 0), direction = Number(e?.direction ?? 0);
  if (![npcId, x, y, z, radiusRaw, direction].every(Number.isFinite)) { stats.skip_bad++; continue; }
  if (npcId < 0 || npcId > 8195 || !validIds.has(Math.trunc(npcId))) { stats.skip_id++; continue; }
  if (z < 0 || z > 3) { stats.skip_plane++; continue; }
  if (x < 0 || y < 0 || x >= 16384 || y >= 16384) { stats.skip_coords++; continue; }
  if (!validRegions.has(regionIdForTile(Math.trunc(x), Math.trunc(y)))) { stats.skip_region++; continue; }

  const rec = {
    facing: toFacing(direction),
    radius: Math.max(0, Math.min(Math.trunc(radiusRaw), 5)),
    id: Math.trunc(npcId),
    position: { x: Math.trunc(x), y: Math.trunc(y), z: Math.trunc(z) },
    description: typeof e?.name === 'string' && e.name.length ? e.name : 'OSRS import',
  };

  const key = `${rec.id}:${rec.position.x}:${rec.position.y}:${rec.position.z}:${rec.facing}:${rec.radius}`;
  if (seen.has(key)) { stats.dupes++; continue; }
  seen.add(key);
  out.push(rec);
  stats.kept++;
}

fs.writeFileSync(outPath, JSON.stringify(out, null, 4), 'utf8');
console.log(`map_index bytes: ${mapIndex.length}`);
console.log(`map regions in index: ${count}`);
console.log(`valid regions: ${validRegions.size}`);
console.log('Rebuilt npc_spawns.json');
for (const [k,v] of Object.entries(stats)) console.log(`${k}: ${v}`);

File: tools/rebuild_npc_spawns_317.py
import json
from pathlib import Path

ROOT = Path(r"d:/CodingProjects/OSRSRSPS")
OSRS_NPCS = ROOT / "osrs_npcs.json"
NPC_DEFS = ROOT / "server/data/definitions/npc_defs.json"
SPAWNS_OUT = ROOT / "server/data/definitions/npc_spawns.json"
CLIP_MAPS = ROOT / "server/data/clipping/maps"


def to_facing(direction: int) -> str:
    # OSRS yaw: 0=S, 512=W, 1024=N, 1536=E.
    sector = ((int(direction) + 128) % 2048) // 256
    return {
        0: "SOUTH",
        1: "SOUTH_WEST",
        2: "WEST",
        3: "NORTH_WEST",
        4: "NORTH",
        5: "NORTH_EAST",
        6: "EAST",
        7: "SOUTH_EAST",
    }[sector]


def region_id(x: int, y: int) -> int:
    return ((x >> 6) << 8) | (y >> 6)


def main() -> None:
    defs = json.loads(NPC_DEFS.read_text(encoding="utf-8"))
    valid_ids = {int(d["id"]) for d in defs if isinstance(d, dict) and "id" in d}

    valid_regions = set()
    for f in CLIP_MAPS.glob("*.dat"):
        try:
            valid_regions.add(int(f.stem))
        except ValueError:
            pass

    entries = json.loads(OSRS_NPCS.read_text(encoding="utf-8"))

    out = []
    seen = set()
    stats = {
        "total": 0,
        "kept": 0,
        "skip_bad": 0,
        "skip_id": 0,
        "skip_region": 0,
        "skip_coords": 0,
        "skip_plane": 0,
        "dupes": 0,
    }

    for e in entries:
        stats["total"] += 1
        try:
            npc_id = int(e.get("id"))
            x = int(e.get("x"))
            y = int(e.get("y"))
            z = int(e.get("z", e.get("p", 0)))
            radius = int(e.get("radius", 0))
            direction = int(e.get("direction", 0))
        except Exception:
            stats["skip_bad"] += 1
            continue

        # 317 client limit and existing defs only.
        if npc_id < 0 or npc_id > 8195 or npc_id not in valid_ids:
            stats["skip_id"] += 1
            continue

        if z < 0 or z > 3:
            stats["skip_plane"] += 1
            continue

        if x < 0 or y < 0 or x >= 16384 or y >= 16384:
            stats["skip_coords"] += 1
            continue

        rid = region_id(x, y)
        if rid not in valid_regions:
            stats["skip_region"] += 1
            continue

        rec = {
            "facing": to_facing(direction),
            "radius": max(0, min(radius, 5)),
            "id": npc_id,
            "position": {"x": x, "y": y, "z": z},
            "description": e.get("name", "OSRS import"),
        }

        key = (rec["id"], x, y, z, rec["facing"], rec["radius"])
        if key in seen:
            stats["dupes"] += 1
            continue

        seen.add(key)
        out.append(rec)
        stats["kept"] += 1

    SPAWNS_OUT.write_text(json.dumps(out, indent=4), encoding="utf-8")

    print("Rebuilt npc_spawns.json")
    for k, v in stats.items():
        print(f"{k}: {v}")


if __name__ == "__main__":
    main()
